compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -4947,6 +4953,28 @@ private STNode parseTerminalExpression(STNode annots, List<STNode> qualifiers, b
         }
     }
 
+    private boolean isValidQueryCtxKeyword(STToken nextToken) {
+        if (!hasAncestorContext(ParserRuleContext.QUERY_EXPRESSION)) {
+            return false;
+        }
```

Comment:
We can do this at lexer level. When we are inside parsing query expression, we capture those tokens as keyword where as in all other places we capture them as identifiers. 
e.g.
```java
    private STNode parseQueryExpression() {
        this.tokenReader.startMode(ParserMode.QUERY);
        // Query expression parsing
        this.tokenReader.endMode();
    }
```
I think that would make many things simple.

Reply:
Updated